File: scripts/create_test_excel.py
#!/usr/bin/env python3
"""
Script to create a custom test Excel document with sample data
"""
import openpyxl
from datetime import datetime, timedelta
import random

def create_test_excel():
    """Create a test Excel file with sample data"""
    
    # Create a new workbook
    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = "Sales Data"
    
    # Add headers
    headers = [
        "Order ID",
        "Customer Name",
        "Product Category",
        "Product Name",
        "Quantity",
        "Unit Price",
        "Total Amount",
        "Order Date",
        "Region",
        "Status",
        "Payment Method",
        "Shipping Cost"
    ]
    
    for col_num, header in enumerate(headers, 1):
        ws.cell(row=1, column=col_num, value=header)
        # Style the header row
        ws.cell(row=1, column=col_num).font = openpyxl.styles.Font(bold=True, color="FFFFFF")
        ws.cell(row=1, column=col_num).fill = openpyxl.styles.PatternFill(
            start_color="4472C4", end_color="4472C4", fill_type="solid"
        )
    
    # Sample data
    categories = ["Electronics", "Clothing", "Home & Garden", "Sports", "Books"]
    products = {
        "Electronics": ["Laptop", "Smartphone", "Tablet", "Headphones", "Camera"],
        "Clothing": ["T-Shirt", "Jeans", "Jacket", "Sneakers", "Watch"],
        "Home & Garden": ["Blender", "Coffee Maker", "Lamp", "Plant Pot", "Tool Set"],
        "Sports": ["Tennis Racket", "Yoga Mat", "Basketball", "Running Shoes", "Gym Bag"],
        "Books": ["Novel", "Textbook", "Magazine", "Comics", "Audiobook"]
    }
    regions = ["North", "South", "East", "West", "Central"]
    statuses = ["Completed", "Pending", "Shipped", "Delivered", "Cancelled"]
    payment_methods = ["Credit Card", "PayPal", "Bank Transfer", "Cash on Delivery"]
    customer_names = [
        "John Smith", "Emily Johnson", "Michael Brown", "Sarah Davis", "David Wilson",
        "Jessica Martinez", "Christopher Lee", "Amanda Taylor", "Daniel Anderson", "Ashley Thomas",
        "Matthew Jackson", "Jennifer White", "Andrew Harris", "Michelle Martin", "Joshua Thompson",
        "Laura Garcia", "James Robinson", "Stephanie Clark", "Robert Rodriguez", "Nicole Lewis"
    ]
    
    # Generate 50 sample records
    base_date = datetime(2024, 1, 1)
    
    for row_num in range(2, 52):  # 50 data rows
        order_id = f"ORD-{row_num:05d}"
        customer_name = random.choice(customer_names)
        category = random.choice(categories)
        product_name = random.choice(products[category])
        quantity = random.randint(1, 10)
        unit_price = round(random.uniform(10.00, 500.00), 2)
        total_amount = round(quantity * unit_price, 2)
        order_date = base_date + timedelta(days=random.randint(0, 364))
        region = random.choice(regions)
        status = random.choice(statuses)
        payment_method = random.choice(payment_methods)
        shipping_cost = round(random.uniform(5.00, 25.00), 2)
        
        ws.cell(row=row_num, column=1, value=order_id)
        ws.cell(row=row_num, column=2, value=customer_name)
        ws.cell(row=row_num, column=3, value=category)
        ws.cell(row=row_num, column=4, value=product_name)
        ws.cell(row=row_num, column=5, value=quantity)
        ws.cell(row=row_num, column=6, value=unit_price)
        ws.cell(row=row_num, column=7, value=total_amount)
        ws.cell(row=row_num, column=8, value=order_date.strftime("%Y-%m-%d"))
        ws.cell(row=row_num, column=9, value=region)
        ws.cell(row=row_num, column=10, value=status)
        ws.cell(row=row_num, column=11, value=payment_method)
        ws.cell(row=row_num, column=12, value=shipping_cost)
    
    # Adjust column widths
    column_widths = {
        'A': 12, 'B': 18, 'C': 16, 'D': 18, 'E': 10, 'F': 12,
        'G': 14, 'H': 12, 'I': 10, 'J': 12, 'K': 18, 'L': 14
    }
    
    for column, width in column_widths.items():
        ws.column_dimensions[column].width = width
    
    # Save the workbook
    filename = "test_data_sales.xlsx"
    wb.save(filename)
    print(f"✓ Test Excel file created: {filename}")
    print(f"  - {len(headers)} columns")
    print(f"  - 50 rows of sample data")
    print(f"  - File location: {filename}")
    
    return filename

if __name__ == "__main__":
    try:
        create_test_excel()
    except ImportError:
        print("❌ Error: openpyxl library not found")
        print("Install it with: pip install openpyxl")
    except Exception as e:
        print(f"❌ Error creating Excel file: {e}")

File: scripts/create_test_excel.js
const XLSX = require('xlsx');
const path = require('path');

function createTestExcel() {
    // Create worksheet
    const ws = XLSX.utils.json_to_sheet([]);
    
    // Add headers
    const headers = [
        "Order ID",
        "Customer Name",
        "Product Category",
        "Product Name",
        "Quantity",
        "Unit Price",
        "Total Amount",
        "Order Date",
        "Region",
        "Status",
        "Payment Method",
        "Shipping Cost"
    ];
    
    // Sample data
    const categories = ["Electronics", "Clothing", "Home & Garden", "Sports", "Books"];
    const products = {
        "Electronics": ["Laptop", "Smartphone", "Tablet", "Headphones", "Camera"],
        "Clothing": ["T-Shirt", "Jeans", "Jacket", "Sneakers", "Watch"],
        "Home & Garden": ["Blender", "Coffee Maker", "Lamp", "Plant Pot", "Tool Set"],
        "Sports": ["Tennis Racket", "Yoga Mat", "Basketball", "Running Shoes", "Gym Bag"],
        "Books": ["Novel", "Textbook", "Magazine", "Comics", "Audiobook"]
    };
    const regions = ["North", "South", "East", "West", "Central"];
    const statuses = ["Completed", "Pending", "Shipped", "Delivered", "Cancelled"];
    const paymentMethods = ["Credit Card", "PayPal", "Bank Transfer", "Cash on Delivery"];
    const customerNames = [
        "John Smith", "Emily Johnson", "Michael Brown", "Sarah Davis", "David Wilson",
        "Jessica Martinez", "Christopher Lee", "Amanda Taylor", "Daniel Anderson", "Ashley Thomas",
        "Matthew Jackson", "Jennifer White", "Andrew Harris", "Michelle Martin", "Joshua Thompson",
        "Laura Garcia", "James Robinson", "Stephanie Clark", "Robert Rodriguez", "Nicole Lewis"
    ];
    
    // Generate data
    const data = [];
    const baseDate = new Date('2024-01-01');
    
    // Add headers as first row
    data.push(headers);
    
    // Generate 50 sample records
    for (let i = 2; i <= 51; i++) {
        const orderId = `ORD-${String(i).padStart(5, '0')}`;
        const customerName = customerNames[Math.floor(Math.random() * customerNames.length)];
        const category = categories[Math.floor(Math.random() * categories.length)];
        const productName = products[category][Math.floor(Math.random() * products[category].length)];
        const quantity = Math.floor(Math.random() * 10) + 1;
        const unitPrice = parseFloat((Math.random() * 490 + 10).toFixed(2));
        const totalAmount = parseFloat((quantity * unitPrice).toFixed(2));
        const orderDate = new Date(baseDate.getTime() + Math.random() * 364 * 24 * 60 * 60 * 1000);
        const region = regions[Math.floor(Math.random() * regions.length)];
        const status = statuses[Math.floor(Math.random() * statuses.length)];
        const paymentMethod = paymentMethods[Math.floor(Math.random() * paymentMethods.length)];
        const shippingCost = parseFloat((Math.random() * 20 + 5).toFixed(2));
        
        data.push([
            orderId,
            customerName,
            category,
            productName,
            quantity,
            unitPrice,
            totalAmount,
            orderDate.toISOString().split('T')[0],
            region,
            status,
            paymentMethod,
            shippingCost
        ]);
    }
    
    // Add data to worksheet
    XLSX.utils.sheet_add_aoa(ws, data);
    
    // Set column widths
    ws['!cols'] = [
        { wch: 12 }, { wch: 18 }, { wch: 16 }, { wch: 18 },
        { wch: 10 }, { wch: 12 }, { wch: 14 }, { wch: 12 },
        { wch: 10 }, { wch: 12 }, { wch: 18 }, { wch: 14 }
    ];
    
    // Create workbook
    const wb = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(wb, ws, "Sales Data");
    
    // Save file
    const filename = path.join(__dirname, 'test_data_sales.xlsx');
    XLSX.writeFile(wb, filename);
    
    console.log('✓ Test Excel file created: test_data_sales.xlsx');
    console.log(`  - ${headers.length} columns`);
    console.log('  - 50 rows of sample data');
    console.log(`  - File location: ${filename}`);
    
    return filename;
}

// Run the function
try {
    createTestExcel();
} catch (error) {
    console.error('❌ Error creating Excel file:', error.message);
    console.error('Make sure xlsx is installed: npm install xlsx');
}
